### ISSUE TITLE
Feature Request: Integrate Google's Perch v2 Bird Vocalization Classifier Model

### ISSUE THREAD

**Issue by tphakala** [tphakala/birdnet-go#1109]
## Feature Description

Add support for Google's Perch v2 bird vocalization classifier model as an alternative or complementary model to the current BirdNET model in birdnet-go.

## Background

Google's Perch v2 is a state-of-the-art bird vocalization classifier that offers several advantages:
- **10,000+ bird species coverage** (vs BirdNET's 6,000)
- **Global training data** from the complete Xeno-Canto corpus
- **High-quality embeddings** (1280-dimensional) for transfer learning
- **Apache 2.0 license** allowing commercial use
- Available via [Kaggle Models](https://www.kaggle.com/models/google/bird-vocalization-classifier/tensorFlow2/perch_v2) and [TensorFlow Hub](https://tfhub.dev/google/bird-vocalization-classifier/1)

### Current Architecture Compatibility
✅ **CGO Integration**: birdnet-go already uses CGO for TFLite integration, making TensorFlow C API integration feasible
✅ **Model Registry**: Existing registry system can be extended for multi-model support
✅ **Audio Pipeline**: Flexible audio processing that can be adapted for different sample rates
✅ **Concurrent Processing**: Existing goroutine-based architecture suitable for multiple models

#### 1. Model Format Differences
- **Current**: TFLite models only
- **Perch v2**: TensorFlow SavedModel format
- **Solution**: Integrate TensorFlow C API via CGO (similar to current TFLite approach)

#### 2. Audio Requirements
- **Current**: 48kHz, 3-second chunks
- **Perch v2**: 32kHz, 5-second chunks
- **Solution**: Add model-specific audio preprocessing pipelines

#### 3. TensorFlow C API Integration
- **Challenge**: Full TensorFlow is heavier than TFLite
- **Options**:
  - Use community bindings like [graft](https://github.com/wamuir/graft) or [tfgo](https://github.com/galeone/tfgo)
  - Direct CGO integration with TensorFlow C API
  - Convert Perch v2 to TFLite format (may lose some capabilities)

#### 4. Output Processing
- **Current**: Species probabilities with sigmoid confidence
- **Perch v2**: 1280-dim embeddings + uncalibrated logits
- **Solution**: Abstract result processing interface

### Phase 1: Architecture Preparation
- [ ] Abstract inference engine interface
- [ ] Create model-agnostic audio preprocessing pipeline
- [ ] Extend model registry for multi-model support
- [ ] Add configuration for model selection

### Phase 2: TensorFlow Integration
- [ ] Integrate TensorFlow C API via CGO
- [ ] Implement SavedModel loader
- [ ] Add 32kHz resampling and 5-second windowing
- [ ] Implement PCEN melspectrogram preprocessing

### Phase 3: Perch v2 Model Support
- [ ] Add Perch v2 model downloader/manager
- [ ] Implement Perch-specific result processing
- [ ] Add 10,000+ species taxonomy support
- [ ] Create embedding extraction mode

### Phase 4: Advanced Features
- [ ] Model comparison/ensemble mode
- [ ] Transfer learning capabilities using embeddings
- [ ] Geographic range filtering with Perch v2
- [ ] Performance benchmarking tools

## Benefits

1. **Expanded Species Coverage**: Support for 10,000+ species globally
2. **Better Accuracy**: Leverage state-of-the-art model performance
3. **Embedding-based Features**: Enable similarity search, clustering, and transfer learning
4. **Model Choice**: Users can select best model for their use case
5. **Future-proofing**: Architecture ready for additional models

## Alternative Approaches

If full TensorFlow integration proves too complex:

1. **TFLite Conversion**: Convert Perch v2 to TFLite (may lose embedding capabilities)
2. **Microservice Architecture**: Run Perch v2 in separate Python service with gRPC/REST API
3. **ONNX Runtime**: Convert to ONNX and use ONNX Runtime Go bindings
4. **Hybrid Approach**: Use TFLite for inference, Python service for embeddings

## Resources

- [Perch v2 on Kaggle](https://www.kaggle.com/models/google/bird-vocalization-classifier/tensorFlow2/perch_v2)
- [Perch GitHub Repository](https://github.com/google-research/perch)
- [TensorFlow C API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h)
- [Graft - TensorFlow Go Bindings](https://github.com/wamuir/graft)

## Questions for Discussion

1. Is there community interest in supporting multiple models?
2. What's the acceptable increase in binary size/dependencies?
3. Should embedding extraction be a priority feature?
4. Preference for TensorFlow C API vs service-based approach?

## Call for Contributors

This is a substantial feature that would benefit from community involvement. If anyone is interested in contributing to this implementation, please comment below. Areas where help would be particularly valuable:

- TensorFlow C API integration expertise
- Model conversion (SavedModel to TFLite/ONNX)
- Audio preprocessing pipeline development
- Testing and benchmarking

---

This feature would significantly enhance birdnet-go's capabilities, making it a more comprehensive tool for bird vocalization analysis with support for cutting-edge models.

**Comment by justinchuby** [comment]
tflite/ONNX conversion: https://huggingface.co/justinchuby/Perch-onnx

**Comment by tphakala** [comment]
> tflite/ONNX conversion: https://huggingface.co/justinchuby/Perch-onnx
Edit. Below is about BirdNET model you converted to ONNX

I was able to interface with ONNX model with Go and got even CUDA GPU acceleration working, this model performs at incredible speeds on a RTX 5080 GPU

=== Performance Metrics ===
Inference time:     4.5907021s
Audio duration:     10799.98s
Real-time ratio:    2352.58x (2352.58s audio processed per second)

That is 3 hours of audio analysed in 4.6 seconds. CPU backed inference takes 50 seconds (both tflite and this ONNX model).

Btw, could you @justinchuby convert range filter meta model to ONNX also?
https://github.com/tphakala/birdnet-go/blob/main/internal/birdnet/data/BirdNET_GLOBAL_6K_V2.4_MData_Model_V2_FP16.tflite

Or share some pointers how I could to that myself.

In any case, thank you! I can now integrate Perch v2 using this same interfacing library.

**Comment by justinchuby** [comment]
You can probably try the tensorflow-onnx project first? I can experiment with it in December

**Comment by justinchuby** [comment]
@tphakala Created the metadata model: https://huggingface.co/justinchuby/BirdNET-onnx/blob/main/birdnet_data_model.onnx

**Comment by tphakala** [comment]
Closing - future research item, Perch v2 integration deferred
